man/man1/pmsnap.1
'\"macro stdmacro
.TH PMSNAP 1 "" "Performance Co-Pilot"
.SH NAME
\f3pmsnap\f1 \- generate performance summary snapshot images
.SH SYNOPSIS
\f3$PCP_BINADM_DIR/pmsnap\f1
[\f3\-NV?\f1]
[\f3\-c\f1 \f2configs\f1]
[\f3\-C\f1 \f2dir\f1]
[\f3\-f\f1 \f2format\f1]
[\f3\-n\f1 \f2names\f1]
[\f3\-o\f1 \f2dir\f1]
.SH DESCRIPTION
.B pmsnap
is a shell script
that is normally run periodically from
.BR crontab (1)
to generate graphic images of
.BR pmchart (1)
performance charts.
These images can be in any of the supported
.B pmchart
formats, including
.IR png ,
.IR bmp ,
and
.IR jpeg ,
and may be incorporated into the content offered by the local web server.
By default
.B pmsnap
generates no textual output unless some error or warning condition is
encountered.
.PP
.B pmsnap
generates images according to its control file,
.B $PCP_PMSNAPCONTROL_PATH
(or
.B dir/control
if the
.B \-C
option is specified),
and uses archives created by
.BR pmlogger (1)
or PCP archive folios created by
.BR pmafm (1)
and
.BR pmlogger_check (1).
Before attempting to configure
.BR pmsnap ,
it is strongly recommended that
.B pmlogger
be configured according to the descriptions in
.BR pmlogger_daily (1),
.BR pmlogger_check (1)
and
.BR pmlogger (1).
.PP
Once
.B pmlogger
has been configured,
it is necessary to configure
.B pmsnap
as follows;
.IP 1.
Edit the control file
.BR $PCP_PMSNAPCONTROL_PATH .
The syntax of this file is described in the comment at the head of the file
and an example is supplied for one and twelve hour "Summary" performance charts
for the local host.
Suitable arguments for
.B pmchart
are also described in the comment.
The user should consult
.B pmchart
for further details.
Note that when
.B pmsnap
is run, it globally substitutes the string
.B LOCALHOSTNAME
with the name of the local host in the control file.
.IP 2.
Test the configuration by running
.ce 1
.BR "$PCP_BINADM_DIR/pmsnap" .
Without any arguments
.B pmsnap
will process every non-comment line in
.BR $PCP_PMSNAPCONTROL_PATH .
The output images will be placed in the files named
in the first field of each line in the control file, with the file format
appended if necessary.
If these file names do not start with
.B /
or
.B .
then they are assumed relative to
.IR dir ,
as specified with the
.B \-o
option.
The default
.I dir
is the current directory.
Note that if
.B pmlogger
has only been recently started (within about the last 15 minutes),
snapshot images may not be produced and no error
messages will be issued - the reason is that
.B pmchart
can not use very short archives
and hence, neither can
.BR pmsnap .
For debugging purposes the
.B \-V
flag should be used.
.IP 3.
Add an appropriate entry for
.B pmsnap
in the
.B root
user's
.BR crontab .
An example is supplied in
.BR $PCP_VAR_DIR/config/pmlogger/crontab .
.IP 4.
Incorporate the
.B pmsnap
images into the local WWW content.
Usually, WWW pages use images that are relative to a particular document root,
so it is often convenient to use the
.B \-o
command line option to specify a sub-directory of the local WWW content,
and then create a web page in this directory that shows the
snapshot images with text and other content appropriate to the local
environment.
.PP
A sample HTML page, suitable for the Summary snapshot may be found in
.BR $PCP_VAR_DIR/config/pmsnap/Summary.html .
.PP
Although
.B pmsnap
attempts to flush
.BR stdio (3)
output buffers in the relevant
.B pmlogger
processes before generating snapshots images,
this may fail for assorted reasons and no error message will be given.
.PP
.B pmsnap
should not be invoked immediately after
.B pmlogger_daily
has rolled the logs because the new archives will be too short
to obtain meaningful results.
Note however that
.B pmsnap
will not report errors from
.B pmchart
about not being able to comply with the
.B \-A
option on very short archives.
In these cases no error will be reported
and no output images will be produced.
.SH OPTIONS
The available command line options are:
.TP 5
.BI \-c " config-pattern"
Only process lines in the control file
which match the
.I config-pattern
regular expression
in the
.B Config
column.
.TP
.BI \-C " dir"
The
.B control
file is located in the directory
.I dir
rather than in the default
.BR $PCP_PMSNAPCONTROL_PATH
location.
.TP
.BI \-f " format"
Set output image format.
The defaults format is
.BR png .
.TP
.BI \-n " name-pattern"
Only process lines in the control file
which match the
.I name-pattern
regular expression (see
.BR egrep (1))
in the
.B Name
column.
.TP
.B \-N
Enable a ``show me'' mode, where the actions are echoed,
but not executed, in the style of ``make \-n''.
Using
.B \-N
in conjunction with
.B \-V
maximizes the diagnostic capabilities for debugging.
.TP
.BI \-o " dir"
The output images having file names which do not start
with
.B /
or
.B .
will be placed in a directory relative to
.IR dir ,
otherwise the output directory
is relative to the current directory (i.e. the default
value for
.I dir
is
.BR ./ ).
Note that
.I dir
must be a writable directory path
and may be on an NFS or CIFS file system.
.TP
.B \-V
Enable verbose mode.
.TP
.B \-?
Display usage message and exit.
.PP
When either
.B \-n
or
.BR \-c
are used,
.B pmsnap
will only process lines in the control file
which match all the supplied patterns.
If no patterns are given,
then all lines will be processed.
These arguments allow multiple entries for
.B pmsnap
in
.B crontab
so that different performance summary images can be generated
at different times or with different frequencies.
.SH FILES
.TP 5
.I $PCP_PMSNAPCONTROL_PATH
\fBpmsnap\fR control file
.TP
.I $PCP_VAR_DIR/config/pmsnap/Summary
summary view for
.B pmchart
.TP
.I $PCP_VAR_DIR/config/pmsnap/Summary.html
sample HTML page for summary snapshot
.TP
.IR $PCP_LOG_DIR/pmlogger/<hostname>/Latest
PCP archive folio for the host
.IR hostname ,
as generated by
.B pmlogger_check
.TP
.I $PCP_VAR_DIR/config/pmlogger/crontab
example
.B crontab
entry
.SH SEE ALSO
.BR cron (1),
.BR crontab (1),
.BR egrep (1),
.BR pmafm (1),
.BR pmchart (1),
.BR pmlc (1),
.BR pmlogger (1)
and
.BR pmlogger_daily (1).

.\" control lines for scripts/man-spell
.\" +ok+ CIFS Config
.\" +ok+ bmp jpeg png {all from file formats}


CHANGELOG
pcp-6.3.6 (17 March 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.4 (13 March 2025)
    - libpcp: fix endian issue in big-endian __pmTimestamp writing
    - pcp-xsos: fine-tune error handling on bad command line options
    - pcp-xsos: single-line pmPrintValue/pminfo values, escape shell chars
    - pcp-atop: correct BPF per-process network metric names
    - pmcd, pmdaproc: allow non-local username connection attributes
    - pmlogger, libpcp: fix handling of dynamic instance domain labels
    - pmrep: use BPF instead of BCC for proc networking metrics
    - pmlogconf: use libbpf over BCC for networking metrics
    - pmlogextract: fix logic around instances labels, allowing set updates
    - pmdalinux: mem.numa.hugepages metrics (per node, per hugepage size)
    - pmdabpftrace: fix for some python3 versions async IO create_task
    - pmdanvidia: if present use dso from the libnvidia-ml runtime package
    - pmdadm, pmieconf: default to enabling the dmthin pmie rules
    - configure.ac: fix autoreconf, add AC_DEFINE_UNQUOTED descriptions
    - rpm builds: enable bpf, bpftrace, bcc on riscv64 builds
    - rpm builds: drop user creation scriptlets for rpm builds
    - deb build: force debian/rules to be recreated each time
    - containers: updates to the archive-analysis container
    - selinux: several policy updates for new AVC denials
    - docs: improve SSL/TLS diagnostic in pcp-verify script

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out

src/pmsnap/pmsnap.sh
#!/bin/sh
#
# Copyright (c) 2008 Aconex.  All Rights Reserved.
# Copyright (c) 1995-2000,2003 Silicon Graphics, Inc.  All Rights Reserved.
# 
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110, USA
# 

. $PCP_DIR/etc/pcp.env

tmp=`mktemp -d "$PCP_TMPFILE_DIR/pmsnap.XXXXXXXXX"` || exit 1
trap "rm -rf $tmp; exit 0" 0 1 2 3 15
prog=`basename $0`

LOCALHOST=`pmhostname`
CONFIGDIR=$PCP_VAR_DIR/config/pmsnap
CONTROL=$CONFIGDIR/control
[ -z "$PCP_PMSNAPCONTROL_PATH" ] || CONTROL="$PCP_PMSNAPCONTROL_PATH"

_usage()
{
    cat << EOF
Usage: $prog [-dNV] [-C dir] [-c regex] [-f format] [-n regex] [-o dir]
    -C    alternate directory for control file(s)
    -c    matches the Config field in the pmsnap control file
    -f    output image format (default png)
    -N    show me, but do not execute commands
    -n    matches the Name field in the pmsnap control file
    -o    default directory for output images (default ".")
    -V    verbose trace of actions (for debugging)
    --    pass all following options directly to pmchart

If no patterns are given then all lines in the pmsnap control file
	$CONTROL
will be processed.  If any patterns are given then only lines which
match all patterns will be processed.

Patterns are full regular expressions, see egrep(1) and may require
appropriate quotes to avoid shell meta character expansion.
EOF
    exit 1
}

_fixpath() 
{
    echo $1 $2 |\
    $PCP_AWK_PROG '{
	if (substr($2, 1, 1) == "/")
	    print $2
	else
	    printf "%s/%s\n", $1, $2
    }'
}

_warning()
{
    echo
    echo "$prog: $CONTROL[$line]:"
    echo "$*" | fmt \
| sed -e '1{
s/^/Warning: /
n
}' -e 's/^/         /'
}

_error()
{
    echo
    echo "$prog: $CONTROL[$line]:"
    echo "$*" | fmt \
| sed -e '1{
s/^/Error: /
n
}' -e 's/^/       /'
}

_debug()
{
    echo
    echo "$prog: $CONTROL[$line]:"
    echo "$*" | fmt \
| sed -e '1{
s/^/Debug: /
n
}' -e 's/^/       /'
}

serverId()
{
    id=42
    while [ -f /tmp/.X$id-lock ]; do
	id=`expr $id + 1`
    done
    echo $id
}

startXvfb()
{
    [ -x /usr/bin/Xvfb ] || return

    geom="2048x2048"
    args="-nolisten tcp -screen 0 ${geom}x8"
    server=`serverId`
    cookie=`mcookie`
    XAUTHORITY=$tmp/xauth xauth add ":$server" . "$cookie" >/dev/null 2>&1
    XAUTHORITY=$tmp/xauth /usr/bin/Xvfb ":$server" $args >/dev/null 2>&1 &
    pid=$!
    pmsleep 0.2
    kill -0 $pid 2>/dev/null || return
    export XAUTHORITY=$tmp/xauth
    export DISPLAY=":$server"
    echo $pid
}

stopXvfb()
{
    pid="$1"
    [ -z "$pid" -o "$pid" -le 0 ] && return
    kill $pid
}

line=0
archives=""
configs='.*'
names='.*'
dir="."
format="png"
verbose=false
passthru=false
version="1.0"

# option parsing
#
SHOWME=false
PMLC="echo flush | pmlc >/dev/null 2>&1"
while getopts a:C:c:f:Nn:o:V-? c
do
    case $c
    in
	C)	CONFIGDIR="$OPTARG"
		;;
	c)	configs="$OPTARG"
		;;
	f)	format="$OPTARG"
		;;
	o)	dir="$OPTARG"
		;;
	N)	SHOWME=true
		PMLC="echo '+ echo flush | pmlc'"
		;;
	n)	names="$OPTARG"
		;;
	V)	verbose=true
		;;
	-)	passthru=true
		;;
	?)	_usage
		;;
    esac
    [ $passthru = true ] && break
done

if [ $passthru = false ]
then
    shift `expr $OPTIND - 1`
    [ $# -ne 0 ] && _usage
fi
commonargs="$commonargs $@"

CONTROL=$CONFIGDIR/control
if [ ! -f "$CONTROL" ]
then
    echo "$prog: Error: cannot find control file \"$CONTROL\""
    exit 1
fi

if [ ! -d "$dir" ]
then
    echo "$prog: Error: directory \"$dir\" does not exist"
    exit 1
fi
cd $dir
dir=`pwd`
$SHOWME && echo "+ cd $dir"

$SHOWME || xvfb=`startXvfb`

# escape / in patterns so $PCP_AWK_PROG is not confused
#
names=`echo "$names" | sed -e 's;/;\\\\/;g'`
configs=`echo "$configs" | sed -e 's;/;\\\\/;g'`

sed -e 's/LOCALHOSTNAME/'$LOCALHOST'/g' < $CONTROL \
| $PCP_AWK_PROG '
/^\#/ || /^\$/ || NF < 4	{
    print
    next
}
{
    if ($1 ~ /'$names'/ && $3 ~ /'$configs'/) {
	printf "%s %s %s '\''", $1, $2, $3
	for (i=4; i<= NF; i++)
	    printf "%s ", $i
	printf "'\''\n"
    }
}' \
| while read aname afolio aview args
do
    line=`expr $line + 1`
    case "$aname"
    in
        \#*|'') # comment or empty
                continue
                ;;

	\$*)	# in-line variable assignment
		$SHOWME && echo "# $aname $afolio $aview $args"
		cmd=`echo "$aname $afolio $aview $args" \
		     | sed -n \
			 -e "/='/s/\(='[^']*'\).*/\1/" \
			 -e '/="/s/\(="[^"]*"\).*/\1/' \
			 -e '/=[^"'"'"']/s/[;&<>|].*$//' \
			 -e '/^\\$[A-Za-z][A-Za-z0-9_]*=/{
s/^\\$//
s/^\([A-Za-z][A-Za-z0-9_]*\)=/export \1; \1=/p
}'`
		if [ -z "$cmd" ]
		then
		    # in-line command, not a variable assignment
		    _warning "in-line command is not a variable assignment, line ignored"
		else
		    case "$cmd"
		    in
			'export PATH;'*)
			    _warning "cannot change \$PATH, line ignored"
			    ;;
			'export IFS;'*)
			    _warning "cannot change \$IFS, line ignored"
			    ;;
			*)
			    $SHOWME && echo "+ $cmd"
			    eval $cmd
			    ;;
		    esac
		fi
		continue
		;;
    esac

    if [ -z "$afolio" -o -z "$aview" -o -z "$args" ]
    then
	_error "insufficient fields in control file record"
	continue
    fi

    # strip first and last single quote added by awk prior to read
    #
    args=`echo $args | sed -e "s/^'//" -e "s/'$//"`

    name=`_fixpath $dir $aname`
    if [ -d $PCP_LOG_DIR/pmlogger ]; then
	folio=`_fixpath $PCP_LOG_DIR/pmlogger $afolio`
    fi
    view=`_fixpath $dir $aview`

    # make sure output directory exists
    #
    outdir=`dirname $name`
    if [ ! -d $outdir ]
    then
	if $SHOWME
	then
	    echo "+ mkdir -p $outdir"
	else
	    mkdir -p $outdir
	    if [ ! -d $outdir ]
	    then
		_error "cannot create directory \"$outdir\" for output file"
	    else
		$verbose && _debug "created image output directory \"$outdir\""
	    fi
	fi
    fi

    if $SHOWME
    then
	:
    else
	if [ ! -w $outdir ]
	then
	    _error "cannot write in directory \"$outdir\" to create output file"
	    exit 1
	fi
    fi

    if $verbose
    then
	echo
	_debug "name=\"$name\""
	_debug "folio=\"$folio\""
	_debug "view=\"$view\""
	_debug "args=\"$args\""
	_debug "commonargs=\"$commonargs\""
    fi

    if [ ! -f "$view" ]
    then
	if [ ! -f $CONFIGDIR/$aview ]
	then
	    if [ ! -f $PCP_VAR_DIR/config/pmchart/$aview ]
	    then
		if [ ! -f $PCP_VAR_DIR/config/kmchart/$aview ]
		then
		    _warning "could not find \"$view\", \"$CONFIGDIR/$aview\", \"$PCP_VAR_DIR/config/pmchart/$aview, or \"$PCP_VAR_DIR/config/kmchart/$aview\""
		    continue
		else
		    view=$PCP_VAR_DIR/config/kmchart/$aview
		fi
	    else
		view=$PCP_VAR_DIR/config/pmchart/$aview
	    fi
	else
	    view=$CONFIGDIR/$aview
	fi
    fi

    if [ ! -f $folio ]
    then
	# oops, no folio
	_error "cannot find archive folio \"$folio\""
	echo "Skipping entry from control file ..."
	continue
    fi

    f=$folio
    $SHOWME && echo "+ pmafm $f selection"
    pmafm "$f" selection >$tmp/out 2>&1

    if [ $? -ne 0 ]
    then
	if [ ! -f $folio.meta ]
	then
	    _warning "\"$folio\" is neither a PCP folio nor a PCP archive"
	    ls -l $folio
	    continue
	else
	    contents=$folio
	fi
    else
	contents=`sed -n -e '/Archive:/s///gp' <$tmp/out`
	if [ -z "$contents" ]
	then
	    _warning "cannot determine archive name from PCP folio \"$folio\""
	    continue
	fi
    fi

    arch=""

    for c in $contents
    do
	if [ -f $c.meta ]
	then
	    host=`pminfo -f -a $c pmcd.pmlogger.host \
	    | sed -e 's/"//g' -e 's/\[//g' \
	    | $PCP_AWK_PROG '$1=="inst" {print $6; exit}'`

	    port=`pminfo -f -a $c pmcd.pmlogger.port \
	    | sed -e 's/"//g' -e 's/\[//g' \
	    | $PCP_AWK_PROG '$1=="inst" {print $6; exit}'`

	    if [ ! -z "$host" -a ! -z "$port" ]
	    then
		echo "$host $port" >> $tmp/flush
	    else
		if $verbose
		then
		    _warning "could not extract host and pmlogger control port from archive \"$c\" ... this archive may not be flushed"
		fi
	    fi

	    if [ -z "$arch" ]
	    then
		arch=$c
	    else
		arch=$arch","$c
	    fi
	else
	    _warning "cannot find or open archive \"$c\""
	fi
    done

    if [ -z "$arch" ]
    then
	_warning "no archives could be found for \"$contents\""
	continue
    fi

    sort -u $tmp/flush | while read host port
    do
	eval $PMLC -h $host -p $port
    done

    if $SHOWME
    then
	echo "+ pmchart -c $view -o $name.$format -a $arch $commonargs $args"
    else
	eval pmchart -c "$view" -o "$tmp/name.$format" -a "$arch" $commonargs $args > $tmp/err 2>&1
	sts=$?

	if [ $sts -eq 0 -a -f $tmp/name.$format ]
	then
	    # success, update the image
	    mv $tmp/name.$format $name.$format
	    $verbose && _debug "created \"$name.$format\""
	    if [ -s "$tmp/err" ]
	    then
		_warning "the output image \"$name.$format\" was created but there were warning messages from pmchart, as follows:"
		cat $tmp/err
		echo "-------------"
	    fi
	elif grep 'too short to satisfy the requested starting alignment' $tmp/err >/dev/null
	then
	    # archive's duration is too short for chart configuration ...
	    # silently ignore this one, as we'll probably succeed next time
	    # or soon thereafter
	    if $verbose
	    then
		_debug "archive \"$arch\" too short for pmchart args \"$args\""
	    fi
	else
	    # failed, remove the image and report the error
	    _warning "the output image \"$name.$format\" was not created. There were error messages from pmchart, as follows:"
	    cat $tmp/err
	    echo "-------------"
	    rm -f $name.$format
	fi
    fi

    rm -f $tmp/err
done

$SHOWME || stopXvfb "$xvfb"

exit $status
